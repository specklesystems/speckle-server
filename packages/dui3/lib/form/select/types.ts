export type ProjectsSelectItemType = { id?: string; name: string }
export type ModelsSelectItemType = { id?: string; name: string }
export type VersionsSelectItemType = {
  id: string
  message: string
  createdAt: Date
  referencedObject: string
  sourceApplication: string
}
