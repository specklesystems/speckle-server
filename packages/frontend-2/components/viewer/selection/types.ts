export type KeyValuePair = {
  key: string
  value: unknown
  units?: string
  type?: string
  innerType?: string | null
  arrayLength?: number | null
  arrayPreview?: string | null
  backendPath?: string
}
