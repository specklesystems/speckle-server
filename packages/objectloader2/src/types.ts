export type CustomLogger = (message?: string, ...optionalParams: unknown[]) => void

export interface Item {
  id: string
  obj: Base
}

export interface Base {
  id: string
  __closure?: string[]
}

export function isString(value: unknown): value is string {
  return typeof value === 'string'
}
