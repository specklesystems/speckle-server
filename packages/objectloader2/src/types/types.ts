export type CustomLogger = (message?: string, ...optionalParams: unknown[]) => void

export type Fetcher = (
  input: string | URL | Request,
  init?: RequestInit
) => Promise<Response>

export interface Item {
  baseId: string
  base: Base
  size?: number
}

export interface Base {
  id: string
  speckle_type: string
  __closure?: Record<string, number>
}

export interface Reference {
  speckle_type: string
  referencedId: string
  __closure?: Record<string, number>
}

export interface DataChunk extends Base {
  data?: Base[]
}

export function isBase(maybeBase?: unknown): maybeBase is Base {
  return (
    maybeBase !== null &&
    typeof maybeBase === 'object' &&
    'id' in maybeBase &&
    typeof maybeBase.id === 'string'
  )
}

export function isReference(maybeRef?: unknown): maybeRef is Reference {
  return (
    maybeRef !== null &&
    typeof maybeRef === 'object' &&
    'referencedId' in maybeRef &&
    typeof maybeRef.referencedId === 'string'
  )
}

export function isScalar(
  value: unknown
): value is string | number | boolean | bigint | symbol | undefined {
  const type = typeof value
  return (
    value === null ||
    type === 'string' ||
    type === 'number' ||
    type === 'boolean' ||
    type === 'bigint' ||
    type === 'symbol' ||
    type === 'undefined'
  )
}

export function take<T>(it: Iterator<T>, count: number): T[] {
  const result: T[] = []
  for (let i = 0; i < count; i++) {
    const itr = it.next()
    if (itr.done) break
    result.push(itr.value)
  }
  return result
}
