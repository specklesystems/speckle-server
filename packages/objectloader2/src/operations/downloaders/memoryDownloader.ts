import Queue from '../../helpers/queue.js'
import { Base, Item } from '../../types/types.js'
import { Downloader } from '../interfaces.js'

export class MemoryDownloader implements Downloader {
  #items: Map<string, Base>
  #rootId: string
  #results?: Queue<Item>

  constructor(rootId: string, items: Map<string, Base>) {
    this.#rootId = rootId
    this.#items = items
  }
  initializePool(params: {
    results: Queue<Item>
    total: number
    maxDownloadBatchWait?: number
  }): void {
    this.#results = params.results
  }
  downloadSingle(): Promise<Item> {
    const root = this.#items.get(this.#rootId)
    if (root) {
      return Promise.resolve({ baseId: this.#rootId, base: root })
    }
    throw new Error('Method not implemented.')
  }
  disposeAsync(): Promise<void> {
    return Promise.resolve()
  }
  add(id: string): void {
    const base = this.#items.get(id)
    if (base) {
      this.#results?.add({ baseId: id, base })
      return
    }
    throw new Error('Method not implemented.')
  }
}
