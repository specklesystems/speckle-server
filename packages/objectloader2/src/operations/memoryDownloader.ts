import AsyncGeneratorQueue from '../helpers/asyncGeneratorQueue.js'
import { Base, Item } from '../types/types.js'
import { Downloader } from './interfaces.js'

export class MemoryDownloader implements Downloader {
  #items: Record<string, Base>
  #rootId: string
  #results?: AsyncGeneratorQueue<Item>

  constructor(
    rootId: string,
    items: Record<string, Base>,
    results?: AsyncGeneratorQueue<Item>
  ) {
    this.#rootId = rootId
    this.#items = items
    this.#results = results
  }
  initializePool(): void {}
  downloadSingle(): Promise<Item> {
    const root = this.#items[this.#rootId]
    if (root) {
      return Promise.resolve({ baseId: this.#rootId, base: root })
    }
    throw new Error('Method not implemented.')
  }
  disposeAsync(): Promise<void> {
    return Promise.resolve()
  }
  add(id: string): void {
    const base = this.#items[id]
    if (base) {
      this.#results?.add({ baseId: id, base })
      return
    }
    throw new Error('Method not implemented.')
  }

  addRange(ids: string[]): void {
    ids.forEach((id) => {
      this.add(id)
    })
  }
}
