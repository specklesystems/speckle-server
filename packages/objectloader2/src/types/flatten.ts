import { Base } from './types.js'

/**
 * Flattens a speckle object. It will ignore arrays, null and undefined valuesm, as well as various 'safe to ignore' speckle properties, such as
 * bbox, __closure, __parents, totalChildrenCount.
 * @param obj object to flatten
 * @returns an object with all its props flattened into `prop.subprop.subsubprop`.
 */
export function flattenBase(obj: Base): Record<string, string | number> {
  const flatten = {} as Record<string, string | number>
  for (const k in obj) {
    if (
      k === 'id' ||
      k === '__closure' ||
      k === '__parents' ||
      k === 'bbox' ||
      k === 'totalChildrenCount'
    ) {
      continue
    }

    const v = obj[k]
    if (v === null || v === undefined || Array.isArray(v)) {
      continue
    }
    if (v.constructor === Object) {
      const flattenProp = flattenBase(v as unknown as Base)
      for (const pk in flattenProp) {
        flatten[k + '.' + pk] = flattenProp[pk]
      }
      continue
    }
    const type = typeof v
    if (type === 'string' || type === 'number') {
      flatten[k] = v as string | number
    } else if (type === 'boolean') {
      flatten[k] = v ? 'true' : 'false' // Convert boolean to string for consistency
    }
  }
  if (obj.id) {
    flatten.id = obj.id
  }
  return flatten
}

export function sententizeBase(obj: Base): string {
  const flattened = flattenBase(obj)
  const propertyStrings = Object.entries(flattened).map(([key, value]) => {
    // Make the key more readable (e.g., "Fire Rating" instead of "FireRating")
    const formattedKey = key.replace(/([A-Z])/g, ' $1').trim()
    return `${formattedKey} is ${value}`
  })
  // Join the individual strings with a comma and a space.
  return `This object has the following properties: ${propertyStrings.join(', ')}.`
}
