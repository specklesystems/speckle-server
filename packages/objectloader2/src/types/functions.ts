import { Base, Reference } from './types.js'

export type CustomLogger = (message?: string, ...optionalParams: unknown[]) => void

export type Fetcher = (
  input: string | URL | Request,
  init?: RequestInit
) => Promise<Response>

export function isBase(maybeBase?: unknown): maybeBase is Base {
  return (
    maybeBase !== null &&
    typeof maybeBase === 'object' &&
    'id' in maybeBase &&
    typeof maybeBase.id === 'string'
  )
}

export function isReference(maybeRef?: unknown): maybeRef is Reference {
  return (
    maybeRef !== null &&
    typeof maybeRef === 'object' &&
    'referencedId' in maybeRef &&
    typeof maybeRef.referencedId === 'string'
  )
}

export function isScalar(
  value: unknown
): value is string | number | boolean | bigint | symbol | undefined {
  const type = typeof value
  return (
    value === null ||
    type === 'string' ||
    type === 'number' ||
    type === 'boolean' ||
    type === 'bigint' ||
    type === 'symbol' ||
    type === 'undefined'
  )
}

export function take<T>(it: Iterator<T>, count: number): T[] {
  const result: T[] = []
  for (let i = 0; i < count; i++) {
    const itr = it.next()
    if (itr.done) break
    result.push(itr.value)
  }
  return result
}

export enum ObjectLoader2Flags {
  DEBUG = 'debug',
  USE_CACHE = 'useCache'
}

const defaultValues: Record<ObjectLoader2Flags, string> = {
  [ObjectLoader2Flags.DEBUG]: 'false',
  [ObjectLoader2Flags.USE_CACHE]: 'true'
}

export function getFeatureFlag(
  paramName: ObjectLoader2Flags,
  useDefault: boolean = true
): string | undefined {
  // Check if the code is running in a browser environment üåê
  const isBrowser =
    typeof window !== 'undefined' && typeof window.document !== 'undefined'

  if (!isBrowser) {
    // If in Node.js or another server environment, return the default
    return useDefault ? defaultValues[paramName] : undefined
  }

  // In a browser, parse the query string
  const params = new URLSearchParams(window.location.search)

  // .get() returns the value, or null if it's not found.
  // The nullish coalescing operator (??) provides the default value
  // if the left-hand side is null or undefined.
  return params.get(paramName) ?? (useDefault ? defaultValues[paramName] : undefined)
}

/**
 * Finds the first index of a "needle" Uint8Array within a "haystack" Uint8Array.
 * @param haystack The larger array to search within.
 * @param needle The smaller array to search for.
 * @param start The index to start searching from. Defaults to 0.
 * @returns The starting index of the needle, or -1 if not found.
 */
export function indexOf(
  haystack: Uint8Array,
  needle: Uint8Array,
  start: number = 0
): number {
  if (needle.length === 0) {
    return 0
  }

  // The last possible starting position for a match
  const limit = haystack.length - needle.length

  for (let i = start; i <= limit; i++) {
    let foundMatch = true
    for (let j = 0; j < needle.length; j++) {
      if (haystack[i + j] !== needle[j]) {
        foundMatch = false
        break // Mismatch, break inner loop
      }
    }
    if (foundMatch) {
      return i // Found a full match at index i
    }
  }

  return -1 // No match found
}
