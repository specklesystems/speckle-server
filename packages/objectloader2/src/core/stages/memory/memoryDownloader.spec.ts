import { describe, it, expect, beforeEach } from 'vitest'
import { MemoryDownloader } from './memoryDownloader.js'
import BufferQueue from '../../../queues/bufferQueue.js'
import Queue from '../../../queues/queue.js'
import { Base, Item } from '../../../types/types.js'

const makeBase = (foo: string): Base => ({ foo } as unknown as Base)

describe('MemoryDownloader', () => {
  let items: Map<string, Base>
  let downloader: MemoryDownloader
  let results: Queue<Item>

  beforeEach(() => {
    items = new Map([
      ['id1', makeBase('foo')],
      ['id2', makeBase('bar')]
    ])
    downloader = new MemoryDownloader('id1', items)
    results = new BufferQueue<Item>()
  })

  it('should download the root item', async () => {
    const item = await downloader.downloadSingle()
    expect(item).toEqual({ baseId: 'id1', base: { foo: 'foo' } })
  })

  it('should throw if root item is missing', async () => {
    const missingDownloader = new MemoryDownloader('missing', items)
    await expect(missingDownloader.downloadSingle()).rejects.toThrow(
      'Method not implemented.'
    )
  })

  it('should add found item to results queue', () => {
    downloader.initializePool({ results, total: 2 })
    downloader.add('id2')
    expect(items).toMatchSnapshot()
  })

  it('should throw if added item is missing', () => {
    downloader.initializePool({ results, total: 2 })
    expect(() => downloader.add('missing')).toThrow()
  })

  it('disposeAsync should resolve', async () => {
    await expect(downloader.disposeAsync()).resolves.not.toThrow()
  })
})
