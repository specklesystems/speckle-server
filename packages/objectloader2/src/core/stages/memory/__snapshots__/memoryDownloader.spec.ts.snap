// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MemoryDownloader > should add found item to results queue 1`] = `
Map {
  "id1" => {
    "foo": "foo",
  },
  "id2" => {
    "foo": "bar",
  },
}
`;
